ozi.cc

```diff
@@ -20,30 +20,63 @@
     along with this program; if not, write to the Free Software
     Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111 USA
 
+    Reference:
+    https://www.oziexplorer4.com/eng/help/fileformats.html
+
+    According to the OZI Explorer developer:
+    "There is no specified character set, it defaults to whatever 8 bit
+     character set "Windows" defaults to - normally CP-1252 but can vary
+     depending on Windows regional settings."
+
+    According to the reference, for some text fields:
+    "comma's not allowed in text fields, character 209 can be used instead
+     and a comma will be substituted."
+    This could work for windows-1252, but not for utf-8.
+    We don't support any special handling for character 209.
+
  */
 
+#include <cctype>               // for tolower
+#include <cmath>                // for lround
+#include <cstdlib>              // for atoi
+
+#include <QtCore/QByteArray>    // for QByteArray
+#include <QtCore/QChar>         // for operator==, QChar
+#include <QtCore/QCharRef>      // for QCharRef
+#include <QtCore/QFile>         // for QFile
+#include <QtCore/QFileInfo>     // for QFileInfo
+#include <QtCore/QFlags>        // for QFlags
+#include <QtCore/QIODevice>     // for operator|, QIODevice::WriteOnly, QIODevice::ReadOnly, QIODevice, QIODevice::OpenModeFlag
+#include <QtCore/QString>       // for QString
+#include <QtCore/QStringList>   // for QStringList
+#include <QtCore/QTextCodec>    // for QTextCodec
+#include <QtCore/QTextStream>   // for QTextStream, operator<<, qSetRealNumberPrecision, QTextStream::FixedNotation
+#include <QtCore/Qt>            // for CaseInsensitive
+#include <QtCore/QtGlobal>      // for qPrintable
+
 #include "defs.h"
-#include "cet_util.h"
-#include "csv_util.h"
-#include "jeeps/gpsmath.h"
-#include <QtCore/QFileInfo>
-#include <cctype>
-#include <cmath>                /* for floor */
-#include <cstdio>
-#include <cstdlib>
+#include "csv_util.h"           // for csv_stringclean
+#include "jeeps/gpsmath.h"      // for GPS_Math_Known_Datum_To_WGS84_M
+#include "src/core/datetime.h"  // for DateTime
+#include "src/core/file.h"      // for File
+
 
 #define MYNAME        "OZI"
 #define BADCHARS	",\r\n"
 #define DAYS_SINCE_1990	25569
 
-typedef struct {
+struct ozi_fsdata {
   format_specific_data fs;
   int fgcolor;
   int bgcolor;
-} ozi_fsdata;
+};
 
+static struct {
+  gpsbabel::File* file{nullptr};
+  QTextStream* stream{nullptr};
+  QTextCodec* codec{nullptr};
+} ozi_file;
```

Comment:
All these could become std::unique_ptr<foo> passed in during construction for more safety. Then when you delete ozi_file, they'd be auto-destructed.
I'm actually OK with this as a stepping stone, though.